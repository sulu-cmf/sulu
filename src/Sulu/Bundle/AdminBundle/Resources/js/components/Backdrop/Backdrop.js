// @flow
import classNames from 'classnames';
import {Portal} from 'react-portal';
import React from 'react';
import backdropStyles from './backdrop.scss';

type Props = {
    open: boolean,
    /** When set to false the backdrop renders transparent. */
    visible: boolean,
    /** If true, the backdrop gets rendered in the placed element and not in the body. */
    local: boolean,
    onClick?: () => void,
    fixed: boolean,
};

export default class Backdrop extends React.PureComponent<Props> {
    static defaultProps = {
        open: true,
        visible: true,
        local: false,
        fixed: true,
    };

    handleClick = () => {
        if (this.props.onClick) {
            this.props.onClick();
        }
    };

    render() {
        const {
            open,
            visible,
            local,
            fixed,
        } = this.props;
        const backdropClass = classNames(
            backdropStyles.backdrop,
            {
                [backdropStyles.visible]: visible,
                [backdropStyles.fixed]: fixed,
            }
        );
        const backdrop = <div onClick={this.handleClick} className={backdropClass} />;

        if (!open) {
            return null;
        }

        if (local) {
            return backdrop;
        }

        return <Portal>{backdrop}</Portal>;
    }
}
