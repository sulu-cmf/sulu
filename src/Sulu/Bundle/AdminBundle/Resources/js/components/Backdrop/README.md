The `Backdrop` component serves as a simple solution to create a backdrop for overlays.

Here is a basic example of the component. The open state of the backdrop is controlled by the `open` property.

```
intialState = {open: false};

<div>
    <button onClick={() => setState({open: true})}>Open Backdrop</button>
    <Backdrop open={!!state.open} onClick={() => setState({open: false})} />
</div>
```

This time the `visible` property is set to false, therefore the backdrop is invisible.

```
intialState = {open: false};

<div>
    <button onClick={() => setState({open: true})}>Open Backdrop</button>
    <Backdrop visible={false} open={!!state.open} onClick={() => setState({open: false})} />
</div>
```
