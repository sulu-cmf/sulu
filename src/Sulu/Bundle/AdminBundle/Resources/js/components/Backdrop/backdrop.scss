@import '../../containers/Application/colors.scss';

$backdropBackground: rgba($dustyGray, .9);

.backdrop {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: transparent;

    &.visible {
        background: $backdropBackground;
    }
}
