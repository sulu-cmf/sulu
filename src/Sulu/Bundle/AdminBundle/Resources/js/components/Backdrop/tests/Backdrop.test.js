// @flow
import {mount, shallow} from 'enzyme';
import React from 'react';
import pretty from 'pretty';
import Backdrop from '../Backdrop';

afterEach(() => {
    if (document.body) {
        document.body.innerHTML = '';
    }
});

test('The component should render in body when open', () => {
    const body = document.body;
    const view = mount(<Backdrop open={true} />).render();
    expect(view.html()).toBe('');
    expect(pretty(body ? body.innerHTML : '')).toMatchSnapshot();
});

test('The component should not render in body when local property is set', () => {
    const view = mount(<Backdrop local={true} />).render();
    expect(view).toMatchSnapshot();
});

test('The component should not render in the body when closed', () => {
    const body = document.body;
    const view = mount(<Backdrop open={false} />).render();
    expect(view.html()).toBe(null);
    expect(body ? body.innerHTML : '').toBe('');
});

test('The component should call a function when clicked', () => {
    const onClickSpy = jest.fn();
    const view = shallow(<Backdrop open={true} onClick={onClickSpy} />);

    expect(onClickSpy).toHaveBeenCalledTimes(0);
    view.find('.backdrop').simulate('click');
    expect(onClickSpy).toHaveBeenCalledTimes(1);
});
