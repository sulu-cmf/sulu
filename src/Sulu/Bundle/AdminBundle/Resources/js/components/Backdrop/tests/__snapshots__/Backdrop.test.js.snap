// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The component should not render in body when local property is set 1`] = `
<div
  class="backdrop visible"
/>
`;

exports[`The component should render in body when open 1`] = `
"<div>
  <div data-reactroot=\\"\\" class=\\"backdrop visible\\"></div>
</div>"
`;
